File: crawl_openclaw.py
"""
OpenClaw 文档爬虫
目标网站: https://docs.openclaw.ai/zh-CN

方案 A: 使用 crawl4ai（推荐，支持 JS 渲染）
方案 B: 使用 requests + BeautifulSoup（轻量备用）

安装依赖:
    pip install crawl4ai beautifulsoup4 requests aiofiles
    crawl4ai-setup  # 安装 Playwright 浏览器（方案 A 需要）
"""

import asyncio
import re
import time
from pathlib import Path
from urllib.parse import urljoin, urlparse

# ─────────────────────────────────────────────
# 公共配置
# ─────────────────────────────────────────────
BASE_URL = "https://docs.openclaw.ai/zh-CN"
DOMAIN   = "docs.openclaw.ai"
OUT_DIR  = Path("openclaw_docs")        # 输出目录（每页一个 .md 文件）
ALL_TXT  = Path("openclaw_all.txt")     # 合并输出（单个大文本）
DELAY    = 0.5                          # 礼貌延迟（秒）


# ══════════════════════════════════════════════
# 方案 A：crawl4ai（推荐）
# 优点：支持 JS 渲染的 SPA 文档站；自动提取 Markdown
# ══════════════════════════════════════════════
async def crawl_with_crawl4ai():
    """使用 crawl4ai 异步爬取所有页面，输出 Markdown 文件"""
    try:
        from crawl4ai import AsyncWebCrawler, BrowserConfig, CrawlerRunConfig
        from crawl4ai.content_filter_strategy import PruningContentFilter
        from crawl4ai.markdown_generation_strategy import DefaultMarkdownGenerator
    except ImportError:
        print("❌ crawl4ai 未安装，请执行: pip install crawl4ai && crawl4ai-setup")
        return

    OUT_DIR.mkdir(exist_ok=True)
    visited: set[str] = set()
    queue:   list[str] = [BASE_URL]
    all_content: list[str] = []

    browser_cfg = BrowserConfig(headless=True, verbose=False)

    # 内容过滤：保留正文，去除导航/页脚噪声
    md_generator = DefaultMarkdownGenerator(
        content_filter=PruningContentFilter(threshold=0.45, threshold_type="fixed")
    )
    run_cfg = CrawlerRunConfig(markdown_generator=md_generator)

    async with AsyncWebCrawler(config=browser_cfg) as crawler:
        while queue:
            url = queue.pop(0)
            if url in visited:
                continue
            visited.add(url)

            print(f"🕷  抓取: {url}")
            try:
                result = await crawler.arun(url=url, config=run_cfg)
            except Exception as e:
                print(f"  ⚠️  跳过 {url}：{e}")
                continue

            if not result.success:
                print(f"  ⚠️  失败: {result.error_message}")
                continue

            # ── 保存单页 Markdown ──
            slug = url_to_slug(url)
            md_path = OUT_DIR / f"{slug}.md"
            md_content = result.markdown_v2.fit_markdown or result.markdown or ""
            md_path.write_text(f"# 来源: {url}\n\n{md_content}", encoding="utf-8")
            all_content.append(f"\n\n{'='*60}\n# {url}\n{'='*60}\n\n{md_content}")
            print(f"  ✅ 已保存 → {md_path}")

            # ── 从原始 HTML 提取新链接 ──
            new_links = extract_links(result.html, url)
            for link in new_links:
                if link not in visited:
                    queue.append(link)

            await asyncio.sleep(DELAY)

    # ── 写入合并文件 ──
    ALL_TXT.write_text("\n".join(all_content), encoding="utf-8")
    print(f"\n🎉 完成！共爬取 {len(visited)} 页")
    print(f"   📁 分页文件 → {OUT_DIR}/")
    print(f"   📄 合并文件 → {ALL_TXT}")


# ══════════════════════════════════════════════
# 方案 B：requests + BeautifulSoup（备用）
# 适用于纯静态 HTML 文档站；无需 Playwright
# ══════════════════════════════════════════════
def crawl_with_requests():
    """使用 requests + BeautifulSoup 同步爬取"""
    try:
        import requests
        from bs4 import BeautifulSoup
    except ImportError:
        print("❌ 请执行: pip install requests beautifulsoup4")
        return

    OUT_DIR.mkdir(exist_ok=True)
    visited: set[str] = set()
    queue:   list[str] = [BASE_URL]
    all_content: list[str] = []

    session = requests.Session()
    session.headers.update({
        "User-Agent": (
            "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) "
            "AppleWebKit/537.36 (KHTML, like Gecko) "
            "Chrome/122.0.0.0 Safari/537.36"
        )
    })

    while queue:
        url = queue.pop(0)
        if url in visited:
            continue
        visited.add(url)

        print(f"🕷  抓取: {url}")
        try:
            resp = session.get(url, timeout=15)
            resp.raise_for_status()
        except Exception as e:
            print(f"  ⚠️  跳过 {url}：{e}")
            continue

        soup = BeautifulSoup(resp.text, "html.parser")

        # ── 提取正文 ──
        # 常见文档框架的正文选择器（Docusaurus / Mintlify / GitBook 等）
        content_selectors = [
            "article",
            "main",
            '[class*="content"]',
            '[class*="markdown"]',
            '[class*="prose"]',
            ".doc-content",
            "#content",
        ]
        body_el = None
        for sel in content_selectors:
            body_el = soup.select_one(sel)
            if body_el:
                break
        if not body_el:
            body_el = soup.body or soup

        # 去除导航、页脚、侧边栏等噪声元素
        for tag in body_el.select("nav, footer, header, aside, script, style, [class*='nav'], [class*='sidebar'], [class*='footer']"):
            tag.decompose()

        page_text = html_to_markdown(body_el)

        # ── 保存单页 .md ──
        slug = url_to_slug(url)
        md_path = OUT_DIR / f"{slug}.md"
        md_path.write_text(f"# 来源: {url}\n\n{page_text}", encoding="utf-8")
        all_content.append(f"\n\n{'='*60}\n# {url}\n{'='*60}\n\n{page_text}")
        print(f"  ✅ 已保存 → {md_path}")

        # ── 发现新链接 ──
        new_links = extract_links(resp.text, url)
        for link in new_links:
            if link not in visited:
                queue.append(link)

        time.sleep(DELAY)

    # ── 写入合并文件 ──
    ALL_TXT.write_text("\n".join(all_content), encoding="utf-8")
    print(f"\n🎉 完成！共爬取 {len(visited)} 页")
    print(f"   📁 分页文件 → {OUT_DIR}/")
    print(f"   📄 合并文件 → {ALL_TXT}")


# ══════════════════════════════════════════════
# 工具函数
# ══════════════════════════════════════════════
def extract_links(html: str, current_url: str) -> list[str]:
    """从 HTML 中提取同域内的文档链接"""
    from bs4 import BeautifulSoup
    soup = BeautifulSoup(html, "html.parser")
    links = []
    for a in soup.find_all("a", href=True):
        href = a["href"].strip()
        if not href or href.startswith(("#", "mailto:", "javascript:")):
            continue
        full = urljoin(current_url, href)
        parsed = urlparse(full)
        # 只保留同域、同路径前缀的链接
        if parsed.netloc == DOMAIN and parsed.path.startswith("/zh-CN"):
            # 去掉 fragment
            clean = full.split("#")[0].rstrip("/")
            if clean:
                links.append(clean)
    return list(dict.fromkeys(links))  # 去重保序


def url_to_slug(url: str) -> str:
    """将 URL 转换为合法文件名"""
    parsed = urlparse(url)
    slug = parsed.path.strip("/").replace("/", "__") or "index"
    slug = re.sub(r"[^\w\-]", "_", slug)
    return slug[:120]  # 限制长度


def html_to_markdown(element) -> str:
    """将 BeautifulSoup 元素简单转换为 Markdown 文本"""
    lines = []
    for tag in element.descendants:
        if not hasattr(tag, "name"):
            continue
        if tag.name in ("h1", "h2", "h3", "h4", "h5", "h6"):
            level = int(tag.name[1])
            lines.append(f"\n{'#' * level} {tag.get_text(strip=True)}\n")
        elif tag.name == "p":
            text = tag.get_text(" ", strip=True)
            if text:
                lines.append(f"\n{text}\n")
        elif tag.name in ("li",):
            text = tag.get_text(" ", strip=True)
            if text:
                lines.append(f"- {text}")
        elif tag.name == "code":
            text = tag.get_text(strip=True)
            if text:
                lines.append(f"`{text}`")
        elif tag.name == "pre":
            text = tag.get_text()
            lines.append(f"\n```\n{text}\n```\n")
    return "\n".join(lines)


# ══════════════════════════════════════════════
# 入口
# ══════════════════════════════════════════════
if __name__ == "__main__":
    import sys

    mode = sys.argv[1] if len(sys.argv) > 1 else "crawl4ai"

    if mode == "requests":
        print("📦 使用方案 B：requests + BeautifulSoup")
        crawl_with_requests()
    else:
        print("📦 使用方案 A：crawl4ai（推荐）")
        asyncio.run(crawl_with_crawl4ai())

File: crawl-openclaw.js
#!/usr/bin/env node
/**
 * OpenClaw Docs Crawler
 * ─────────────────────────────────────────────────────────────────
 * Crawls https://docs.openclaw.ai/zh-CN and saves every page as a
 * Markdown file, preserving the original URL directory structure.
 *
 * Usage:
 *   node crawl-openclaw.js
 *
 * Install dependencies first:
 *   npm install axios cheerio turndown robots-parser p-queue chalk
 * ─────────────────────────────────────────────────────────────────
 */

"use strict";

const axios        = require("axios");
const cheerio      = require("cheerio");
const TurndownService = require("turndown");
const RobotsParser = require("robots-parser");
const PQueue       = require("p-queue").default ?? require("p-queue");
const fs           = require("fs");
const path         = require("path");
const { URL }      = require("url");

// ── try to load chalk (optional, graceful fallback) ──────────────
let chalk;
try { chalk = require("chalk"); }
catch { chalk = { green: s=>s, yellow: s=>s, red: s=>s, cyan: s=>s, gray: s=>s, bold: s=>s }; }

// ═══════════════════════════════════════════════════════════════════
// CONFIG
// ═══════════════════════════════════════════════════════════════════
const CONFIG = {
  baseUrl:    "https://docs.openclaw.ai/zh-CN",
  origin:     "https://docs.openclaw.ai",
  pathPrefix: "/zh-CN",                                   // only follow links under this prefix
  outDir:     "/home/yxs/.openclaw/workspace/docs-openclaw",
  concurrency: 3,                                         // parallel requests
  delayMs:    400,                                        // ms between requests (rate-limit)
  timeout:    15_000,                                     // request timeout ms
  maxRetries: 3,
  userAgent:  "Mozilla/5.0 (compatible; DocsBot/1.0; +https://github.com/local)",
  respectRobots: true,
};

// ═══════════════════════════════════════════════════════════════════
// GLOBALS
// ═══════════════════════════════════════════════════════════════════
const visited    = new Set();          // normalised URLs already processed
const discovered = new Set();          // all URLs ever enqueued
let   robotsRules = null;              // robots-parser instance

// Turndown: HTML → Markdown
const td = new TurndownService({
  headingStyle:   "atx",
  codeBlockStyle: "fenced",
  bulletListMarker: "-",
});
// Keep <table> as markdown table (via gfm-like conversion)
td.addRule("table", {
  filter: "table",
  replacement(_, node) {
    // Let cheerio serialise then re-parse so we can walk rows
    return "\n\n" + tableToMd(node) + "\n\n";
  },
});

// ═══════════════════════════════════════════════════════════════════
// ROBOTS.TXT
// ═══════════════════════════════════════════════════════════════════
async function loadRobots() {
  try {
    const res = await axios.get(`${CONFIG.origin}/robots.txt`, {
      timeout: 8000,
      headers: { "User-Agent": CONFIG.userAgent },
      validateStatus: () => true,
    });
    if (res.status === 200) {
      robotsRules = RobotsParser(`${CONFIG.origin}/robots.txt`, res.data);
      console.log(chalk.gray("  ✔ robots.txt loaded"));
    } else {
      console.log(chalk.gray("  ℹ  No robots.txt found (HTTP " + res.status + ") — continuing"));
    }
  } catch {
    console.log(chalk.gray("  ℹ  robots.txt unreachable — continuing"));
  }
}

function isAllowed(url) {
  if (!CONFIG.respectRobots || !robotsRules) return true;
  return robotsRules.isAllowed(url, CONFIG.userAgent) !== false;
}

// ═══════════════════════════════════════════════════════════════════
// HTTP  (with retry + exponential back-off)
// ═══════════════════════════════════════════════════════════════════
async function fetchHtml(url, attempt = 1) {
  try {
    const res = await axios.get(url, {
      timeout: CONFIG.timeout,
      headers: {
        "User-Agent": CONFIG.userAgent,
        "Accept":     "text/html,application/xhtml+xml,*/*",
        "Accept-Language": "zh-CN,zh;q=0.9,en;q=0.8",
      },
      maxRedirects: 5,
    });
    return res.data;
  } catch (err) {
    if (attempt < CONFIG.maxRetries) {
      const wait = attempt * 1500;
      console.warn(chalk.yellow(`    ⚠  Retry ${attempt}/${CONFIG.maxRetries} for ${url} (${wait}ms)`));
      await sleep(wait);
      return fetchHtml(url, attempt + 1);
    }
    throw err;
  }
}

// ═══════════════════════════════════════════════════════════════════
// LINK EXTRACTION
// ═══════════════════════════════════════════════════════════════════
function extractLinks($, currentUrl) {
  const links = [];
  $("a[href]").each((_, el) => {
    const href = $(el).attr("href").trim();
    if (!href || href.startsWith("mailto:") || href.startsWith("javascript:")) return;
    try {
      const resolved = new URL(href, currentUrl);
      resolved.hash = "";                       // strip fragment
      const full = resolved.href.replace(/\/$/, ""); // strip trailing slash
      const parsed = new URL(full);
      if (
        parsed.hostname === new URL(CONFIG.origin).hostname &&
        parsed.pathname.startsWith(CONFIG.pathPrefix)
      ) {
        links.push(full);
      }
    } catch { /* ignore malformed */ }
  });
  return [...new Set(links)];
}

// ═══════════════════════════════════════════════════════════════════
// CONTENT EXTRACTION
// ═══════════════════════════════════════════════════════════════════
// Priority selectors for main content (Docusaurus / Mintlify / GitBook)
const CONTENT_SELECTORS = [
  "article",
  "main .markdown",
  "main [class*='content']",
  "main [class*='prose']",
  ".doc-content",
  "[class*='docContent']",
  "[class*='articleBody']",
  "main",
  "#content",
  ".content",
];

// Noise selectors to remove before conversion
const NOISE_SELECTORS = [
  "nav", "header", "footer", "aside",
  "[class*='sidebar']", "[class*='toc']",
  "[class*='breadcrumb']", "[class*='pagination']",
  "[class*='Sidebar']", "[class*='NavBar']",
  "[class*='footer']", "[class*='Footer']",
  "[class*='cookie']", "[class*='banner']",
  "script", "style", "noscript",
];

function extractMainContent($) {
  // Remove noise first
  NOISE_SELECTORS.forEach(sel => $(sel).remove());

  for (const sel of CONTENT_SELECTORS) {
    const el = $(sel).first();
    if (el.length && el.text().trim().length > 100) {
      return el.html() || "";
    }
  }
  return $("body").html() || "";
}

function extractTitle($) {
  return (
    $("h1").first().text().trim() ||
    $("title").first().text().trim().split("|")[0].trim() ||
    "Untitled"
  );
}

// ═══════════════════════════════════════════════════════════════════
// URL → FILE PATH
// ═══════════════════════════════════════════════════════════════════
function urlToFilePath(url) {
  const parsed = new URL(url);
  let   segments = parsed.pathname
        .replace(CONFIG.pathPrefix, "")
        .split("/")
        .filter(Boolean)
        .map(s => slugify(s));

  if (segments.length === 0) {
    return path.join(CONFIG.outDir, "index.md");
  }
  // Last segment becomes the filename; rest become directories
  const last = segments.pop();
  const dir  = path.join(CONFIG.outDir, ...segments);
  return path.join(dir, `${last}.md`);
}

function slugify(str) {
  return str
    .toLowerCase()
    .replace(/[^a-z0-9\u4e00-\u9fff._-]/g, "-")  // keep CJK, alphanumeric, . _ -
    .replace(/-{2,}/g, "-")
    .replace(/^-|-$/g, "");
}

// ═══════════════════════════════════════════════════════════════════
// TABLE HELPER (basic HTML table → GFM)
// ═══════════════════════════════════════════════════════════════════
function tableToMd(node) {
  try {
    const $ = cheerio.load(node.toString ? node.toString() : "");
    const rows = [];
    $("tr").each((_, tr) => {
      const cells = [];
      $(tr).find("th,td").each((_, td) => {
        cells.push($(td).text().replace(/\|/g, "\\|").replace(/\n/g, " ").trim());
      });
      rows.push("| " + cells.join(" | ") + " |");
    });
    if (rows.length < 1) return "";
    const sep = "| " + rows[0].slice(2, -2).split(" | ").map(() => "---").join(" | ") + " |";
    rows.splice(1, 0, sep);
    return rows.join("\n");
  } catch { return ""; }
}

// ═══════════════════════════════════════════════════════════════════
// SAVE MARKDOWN
// ═══════════════════════════════════════════════════════════════════
function saveMd(filePath, title, url, markdown) {
  const dir = path.dirname(filePath);
  fs.mkdirSync(dir, { recursive: true });

  const frontmatter = [
    "---",
    `title: "${title.replace(/"/g, '\\"')}"`,
    `source: "${url}"`,
    `crawled_at: "${new Date().toISOString()}"`,
    "---",
    "",
  ].join("\n");

  fs.writeFileSync(filePath, frontmatter + markdown, "utf-8");
}

// ═══════════════════════════════════════════════════════════════════
// PROCESS ONE PAGE
// ═══════════════════════════════════════════════════════════════════
async function processPage(url, queue) {
  if (visited.has(url)) return;
  visited.add(url);

  if (!isAllowed(url)) {
    console.log(chalk.yellow(`  🚫 Blocked by robots.txt: ${url}`));
    return;
  }

  let html;
  try {
    html = await fetchHtml(url);
  } catch (err) {
    console.error(chalk.red(`  ✗ Failed: ${url} — ${err.message}`));
    return;
  }

  const $     = cheerio.load(html);
  const title = extractTitle($);
  const contentHtml = extractMainContent($);

  // Convert HTML → Markdown
  let markdown = "";
  try {
    markdown = td.turndown(contentHtml);
    // Clean up excessive blank lines
    markdown = markdown.replace(/\n{3,}/g, "\n\n").trim();
  } catch (err) {
    console.warn(chalk.yellow(`  ⚠  Turndown error on ${url}: ${err.message}`));
    // Fallback: plain text
    markdown = cheerio.load(contentHtml).text().trim();
  }

  // Save file
  const filePath = urlToFilePath(url);
  saveMd(filePath, title, url, markdown);
  console.log(chalk.green(`  ✔ [${visited.size}] ${title}`));
  console.log(chalk.gray(`       → ${filePath}`));

  // Enqueue new links
  const links = extractLinks($, url);
  let newCount = 0;
  for (const link of links) {
    if (!discovered.has(link)) {
      discovered.add(link);
      newCount++;
      queue.add(() => processPage(link, queue));
    }
  }
  if (newCount > 0) {
    console.log(chalk.gray(`       + ${newCount} new link(s) discovered`));
  }

  // Rate-limit courtesy delay
  await sleep(CONFIG.delayMs);
}

// ═══════════════════════════════════════════════════════════════════
// MAIN
// ═══════════════════════════════════════════════════════════════════
async function main() {
  console.log(chalk.bold.cyan("\n╔══════════════════════════════════════════╗"));
  console.log(chalk.bold.cyan("║       OpenClaw Docs Crawler v1.0         ║"));
  console.log(chalk.bold.cyan("╚══════════════════════════════════════════╝\n"));
  console.log(chalk.cyan(`  Base URL : ${CONFIG.baseUrl}`));
  console.log(chalk.cyan(`  Out Dir  : ${CONFIG.outDir}`));
  console.log(chalk.cyan(`  Workers  : ${CONFIG.concurrency}`));
  console.log(chalk.cyan(`  Delay    : ${CONFIG.delayMs}ms\n`));

  // Ensure output directory exists
  fs.mkdirSync(CONFIG.outDir, { recursive: true });

  // Load robots.txt
  await loadRobots();

  // Bounded concurrency queue
  const queue = new PQueue({ concurrency: CONFIG.concurrency });

  // Seed with base URL
  const seed = CONFIG.baseUrl.replace(/\/$/, "");
  discovered.add(seed);
  queue.add(() => processPage(seed, queue));

  // Wait for all tasks to complete
  await queue.onIdle();

  console.log(chalk.bold.green(`\n✅ Crawl complete!`));
  console.log(chalk.green(`   Pages saved : ${visited.size}`));
  console.log(chalk.green(`   Output dir  : ${CONFIG.outDir}`));
}

// ═══════════════════════════════════════════════════════════════════
// UTILS
// ═══════════════════════════════════════════════════════════════════
function sleep(ms) { return new Promise(r => setTimeout(r, ms)); }

// ─── Run ──────────────────────────────────────────────────────────
main().catch(err => {
  console.error(chalk.red("\n💥 Fatal error:"), err);
  process.exit(1);
});
